sklearn/utils/validation.py

```diff
@@ -267,11 +270,19 @@ def check_array(array, accept_sparse=None, dtype=None, order=None, copy=False,
         accept_sparse = [accept_sparse]
 
     if sp.issparse(array):
+        if dtype == "numeric":
+            dtype = None
         array = _ensure_sparse_format(array, accept_sparse, dtype, order,
                                       copy, force_all_finite)
     else:
         if ensure_2d:
             array = np.atleast_2d(array)
+        if dtype == "numeric":
+            if getattr(array, "dtype", None) is np.dtype(object):
+                # if input is object, convert to float.
+                dtype = np.float
```

Comment:
I would like to ban the use of `np.float` which is just an alias for Python `float`. Instead we should use the more explicit and platform-independent `np.float64`.